pkg/cmd/openshift-operators/apiserver-operator/controller/object_merger.go

```diff
@@ -0,0 +1,325 @@
+package controller
+
+import (
+	"reflect"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiregistrationv1beta1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1"
+)
+
+func ensureAPIService(modified *bool, existing *apiregistrationv1beta1.APIService, required apiregistrationv1beta1.APIService) {
+	ensureObjectMeta(modified, &existing.ObjectMeta, required.ObjectMeta)
+
+	// we stomp everything
+	if !equality.Semantic.DeepEqual(existing.Spec, required.Spec) {
+		*modified = true
+		existing.Spec = required.Spec
+	}
+}
+
+func ensureObjectMeta(modified *bool, existing *metav1.ObjectMeta, required metav1.ObjectMeta) {
+	setString(modified, &existing.Namespace, required.Namespace)
+	setString(modified, &existing.Name, required.Name)
+	mergeMap(modified, &existing.Labels, required.Labels)
+	mergeMap(modified, &existing.Annotations, required.Annotations)
+}
+
+func ensureService(modified *bool, existing *corev1.Service, required corev1.Service) {
+	ensureObjectMeta(modified, &existing.ObjectMeta, required.ObjectMeta)
+
+	if !reflect.DeepEqual(existing.Spec.Selector, required.Spec.Selector) {
+		*modified = true
+		existing.Spec.Selector = required.Spec.Selector
+	}
+
+	// any port we specify, we require
+	for _, required := range required.Spec.Ports {
+		var existingCurr *corev1.ServicePort
+		for j, curr := range existing.Spec.Ports {
+			if curr.Name == required.Name {
+				existingCurr = &existing.Spec.Ports[j]
+				break
+			}
+		}
+		if existingCurr == nil {
+			*modified = true
+			existing.Spec.Ports = append(existing.Spec.Ports, corev1.ServicePort{})
+			existingCurr = &existing.Spec.Ports[len(existing.Spec.Ports)-1]
+		}
+		ensureServicePort(modified, existingCurr, required)
+	}
+}
+
+func ensureDaemonSet(modified *bool, existing *appsv1.DaemonSet, required appsv1.DaemonSet) {
+	ensureObjectMeta(modified, &existing.ObjectMeta, required.ObjectMeta)
+
+	if existing.Spec.Selector == nil {
+		*modified = true
+		existing.Spec.Selector = required.Spec.Selector
+	}
+	if !reflect.DeepEqual(existing.Spec.Selector, required.Spec.Selector) {
+		*modified = true
+		existing.Spec.Selector = required.Spec.Selector
+	}
+
+	ensurePodTemplateSpec(modified, &existing.Spec.Template, required.Spec.Template)
+}
+
+func ensurePodTemplateSpec(modified *bool, existing *corev1.PodTemplateSpec, required corev1.PodTemplateSpec) {
+	ensureObjectMeta(modified, &existing.ObjectMeta, required.ObjectMeta)
+
+	ensurePodSpec(modified, &existing.Spec, required.Spec)
+}
+
+func ensurePodSpec(modified *bool, existing *corev1.PodSpec, required corev1.PodSpec) {
+	// any container we specify, we require
+	for _, required := range required.Containers {
```

Comment:
This is sort of a core principle.  Do we want to attempt to use a generic merging logic based on metadata or do we want to explicitly say, "I stomp at this level based on these matching keys that I identify like this"?

It seems a little weird, but given minimally specified objects, this gives a lot of confidence that nothing surprising is happening.  It can't accidentally work.  Tectonic uses similar, slightly less specific logic initially and then uses a patch (if I'm reading it right).  This does the whole merge here and then updates or creates.